File: 01_Arrays and Two Pointers/10_MoveZeroes/10_MoveZeroes.js
// https://leetcode.com/problems/move-zeroes/

function moveZeroes(nums) {
    let i = 0, count = 0;
    while (i < nums.length) {
        if (nums[i] == 0) {
            i++;
        } else {
            [nums[count], nums[i]] = [nums[i], nums[count]];
            i++;
            count++;

        }
    }
    
    return nums;
}

console.log(moveZeroes([0,1,0,3,12]));

File: 01_Arrays and Two Pointers/10_MoveZeroes/10_MoveZeroes.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;


class Solution {
public:
    void moveZeroes(vector<int> &nums) {
        int i = 0, count = 0;
        while (i < nums.size()) {
            if (nums[i] == 0) {
                i++;
            } else {
                swap(nums[i], nums[count]);
                i++;
                count++;
            }
        }
    }
};


int main() {
    Solution s;
    vector<int> arg = {0,1,0,3,12};
    s.moveZeroes(arg);
    for (auto x: arg) {
        cout << x << ' ';
    }
    return 0;
}